ndb_api/clone_helpers.go

```diff
@@ -220,65 +220,66 @@ func setCloneNodesParameters(req *DatabaseCloneRequest, database DatabaseInterfa
 	networkProfileId := req.Nodes[0].NetworkProfileId
 	serverTimeZone := req.Nodes[0].NewDbServerTimeZone
 
-	// Clear the original req.Nodes array
+	// Convert database.Instance.Nodes to the common type Nodes
 	req.Nodes = []Node{}
-
-	// Create node object for HA Proxy
-	for i := 0; i < 2; i++ {
-		// Hard coding the HA Proxy properties
-		props := make([]map[string]string, 1)
-		props[0] = map[string]string{
-			"name":  "node_type",
-			"value": "haproxy",
-		}
-		req.Nodes = append(req.Nodes, Node{
-			Properties:  props,
-			VmName:      database.GetName() + "_haproxy" + strconv.Itoa(i),
-			NxClusterId: database.GetClusterId(),
-		})
-	}
-
-	// Create node object for Database Instances
-	for i := 0; i < 3; i++ {
-		// Hard coding the DB properties
-		props := make([]map[string]string, 4)
-		props[0] = map[string]string{
-			"name":  "role",
-			"value": "Secondary",
-		}
-		// 1st node will be the primary node
-		if i == 0 {
-			props[0]["value"] = "Primary"
-		}
-		props[1] = map[string]string{
-			"name":  "failover_mode",
-			"value": "Automatic",
-		}
-		props[2] = map[string]string{
-			"name":  "node_type",
-			"value": "database",
+	for _, node := range database.GetInstanceNodes() {
+		built := Node{}
+		if node.Properties.NodeType == "haproxy" {
+			built = buildHAProxyNode(req, node, database.GetClusterId())
+		} else {
+			built = buildDatabaseNode(req, node, computeProfileId, networkProfileId, serverTimeZone, database.GetClusterId())
 		}
-		props[3] = map[string]string{
-			"name":  "remote_archive_destination",
-			"value": "",
-		}
-		req.Nodes = append(req.Nodes, Node{
-			ComputeProfileId:    computeProfileId,
-			NetworkProfileId:    networkProfileId,
-			NewDbServerTimeZone: serverTimeZone,
-			Properties:          props,
-			VmName:              database.GetName() + "-" + strconv.Itoa(i),
-			NxClusterId:         database.GetClusterId(),
-		})
+
+		req.Nodes = append(req.Nodes, built)
+	}
+}
+
+func buildHAProxyNode(req *DatabaseCloneRequest, node *v1alpha1.Node, clusterId string) Node {
```

Comment:
I don't like the mismatch between the database Node struct and the common Node struct. Do you all think it would be worth it to merge them into a single struct?

Reply:
The common node contains quite a few fields that they don't want the user to be able to change using the CR and using the common node would allow the user to specify values for those fields and we would just have to ignore them or throw an error